lib/vagrant-parallels/action/box_register.rb

```diff
@@ -0,0 +1,105 @@
+require 'log4r'
+
+#require 'digest/md5'
+
+module VagrantPlugins
+  module Parallels
+    module Action
+      class BoxRegister
+        def initialize(app, env)
+          @app = app
+          @logger = Log4r::Logger.new('vagrant_parallels::action::box_register')
+        end
+
+        def call(env)
+          # If we don't have a box, nothing to do
+          if !env[:machine].box
+            return @app.call(env)
+          end
+
+          # Do the register while locked so that nobody else register
+          # a box at the same time.
+          lock_key = Digest::MD5.hexdigest(env[:machine].box.name)
+          env[:machine].env.lock(lock_key, retry: true) do
```

Comment:
There is an issue. `Environment#lock` provides process-level locks only. I've asked Mitchell about it: https://github.com/mitchellh/vagrant/pull/6378/files#r43583268

But it seems like we should add a class-wide mutex here.